Remotion/Data/DomainObjects.UnitTests/Persistence/Rdbms/SqlServer/DbCommandBuilders/SqlQueryDbCommandBuilderTests.cs

```diff
@@ -0,0 +1,157 @@
+using System.Data.SqlClient;
+using Moq;
+using NUnit.Framework;
+using Remotion.Data.DomainObjects.Persistence.Rdbms;
+using Remotion.Data.DomainObjects.Persistence.Rdbms.Model;
+using Remotion.Data.DomainObjects.Persistence.Rdbms.Parameters;
+using Remotion.Data.DomainObjects.Persistence.Rdbms.SqlServer;
+using Remotion.Data.DomainObjects.Persistence.Rdbms.SqlServer.DbCommandBuilders;
+using Remotion.Data.DomainObjects.Persistence.Rdbms.SqlServer.Parameters;
+using Remotion.Data.DomainObjects.Queries;
+using Remotion.Data.DomainObjects.UnitTests.Factories;
+using Remotion.Data.DomainObjects.UnitTests.Persistence.Rdbms.Parameters;
+using Remotion.Utilities;
+
+namespace Remotion.Data.DomainObjects.UnitTests.Persistence.Rdbms.SqlServer.DbCommandBuilders;
+
+[TestFixture]
+public class SqlQueryDbCommandBuilderTests
+{
+  [Test]
+  public void Create_WithTableValuedParameter_UsesTempTable ()
+  {
+    var records = new[]
+                  {
+                      new Guid("280D68FF-BEA9-47CE-BAEC-44569517829D"),
+                      new Guid("6AE44A0F-2345-49AA-95C0-ABB8E968E5CA"),
+                      new Guid("4F5D0D1E-EF63-4D3A-9519-585FF464477B")
+                  };
+
+    // set up a table type with a single Guid column
+    var storageTypeInformation = StorageTypeInformationObjectMother.CreateUniqueIdentifierStorageTypeInformation();
+    var storagePropertyDefinition = new SimpleStoragePropertyDefinition(typeof(Guid), new ColumnDefinition("Value", storageTypeInformation, false));
+    var tableTypeDefinition = new TableTypeDefinition(new EntityNameDefinition(null, "Test"), [ storagePropertyDefinition ], Array.Empty<ITableConstraintDefinition>());
+
+    // set up a record definition with a single property that uses an integer input as index to the records array
+    var recordPropertyDefinition = new FakeRecordPropertyDefinition(storagePropertyDefinition, o => o);
+    var recordDefinition = new RecordDefinition("Test", tableTypeDefinition, [ recordPropertyDefinition ]);
+
+    var parameterDefinition = new SqlTableValuedDataParameterDefinition(recordDefinition);
+    var commandBuilder = new SqlDbQueryCommandBuilder(
+        "SELECT * FROM [MyTable] WHERE MyRefID IN (SELECT [Value] FROM @tvp)",
+        [new QueryParameterWithDataParameterDefinition(new QueryParameter("@tvp", records), parameterDefinition)],
+        new SqlDialect());
+
+    var commandFactoryStub = new Mock<IDbCommandFactory>();
+    commandFactoryStub.Setup(stub => stub.CreateDbCommand()).Returns(new SqlCommand());
+
+    var result = commandBuilder.Create(commandFactoryStub.Object);
+
+    Assert.That(
+        result.CommandText,
+        Is.EqualTo("""
+                   SELECT [Value] INTO [#@tvp] FROM @tvp;
+                   CREATE NONCLUSTERED INDEX [IX_@tvp] ON [#@tvp] ([Value]);
+
+                   SELECT * FROM [MyTable] WHERE MyRefID IN (SELECT [Value] FROM [#@tvp])
+
+                   DROP TABLE [#@tvp];
+
+                   """));
```

Comment:
loopy stuff feels like a case where we should test the zero-items and the more-than-one-items cases. the zeri-items case might be overkill when testing, although it's also rather simple.

And we should test that we only do this for tvp parameters.